src/Config.nebel.old.txt

useConfig Default

// FRONT---web--------> WEBPORT [WebAngular ]---( liteServer / docker )
//                \   (if internal)
//                 \
// FRONT---comands----> PROXYPORT [Proxy ] ----> NEBELPORT [NEBELDOCKER]

Config: Default
    version 'Middlenebel v0.1.3-alpha Nebel-Docker'
    webPort 3000
    proxyPort 8081
    nebelPort 8080
    front internal //off //lite //docker
    frontName 'my-angular-app'
.
Config: DEVELOP
    version 'Middlenebel v0.1.3-alpha Nebel-Docker - DEVELOP'
    proxyPort 8080
    front off
.
Config: NotParse
    version 'Middlenebel v0.1.3-alpha Nebel-Docker - Debuging server'
    disableParse true
    dumpConfig true
    serverPort 8080
.
Config: NotServer
    version 'Middlenebel v0.1.3-alpha Nebel-Docker - Debuging parser'
    disableServer true
    dumpConfig true
    serverPort 8080
.

doc/02-CPP-Environment.md

# Start Front

*ng serve*
.../middlenebel-front$ ng serve --port 8081

*lite-server*
.../middlenebel-front$ npm run dev
Note: the port is configured in package.json in definition of 'dev'

*Docker*
docker run -p 8081:80 my-angular-app


# Configure IDE for debug

https://github.com/microsoft/vscode-makefile-tools/issues/212

https://github.com/bloominstituteoftechnology/CS-Wiki/wiki/C-and-Cpp-Debugging-in-VS-Code

# Installing MySQL C++ Connector
https://www.codingfriends.com/index.php/2010/02/17/mysql-connection-example/comment-page-1/
https://askubuntu.com/questions/165868/installing-mysql-connector-c
https://dev.mysql.com/doc/connector-cpp/8.3/en/connector-cpp-installation-binary.html#connector-cpp-installation-binary-linux
https://aaronxu17.github.io/blog/install-mysql-connector/

apt-get install libmysqlcppconn-dev

git clone https://github.com/mysql/mysql-connector-cpp.git

#  Util Commands

**Find library installed (as mysql)** apt list --installed |grep mysql

libmysqlcppconn-dev/stable,now 1.1.12-4 amd64 [installed]
libmysqlcppconn7v5/stable,now 1.1.12-4 amd64 [installed,automatic]
mysql-common/stable,now 5.8+1.1.0 all [installed,automatic]

**Find installation location of package** dpkg -L libmysqlcppconn-dev

/usr/include/mysql_connection.h
/usr/include/mysql_driver.h
/usr/include/mysql_error.h
[...]
/usr/lib/x86_64-linux-gnu/libmysqlcppconn.a
/usr/lib/x86_64-linux-gnu/libmysqlcppconn.so


src/inc/Config.hpp
/*
 * Copyright (c) 2024, David Hernández Tejada
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 * * Redistributions of source code must retain the above copyright
 *   notice, this list of conditions and the following disclaimer.
 * * Redistributions in binary form must reproduce the above
 *   copyright notice, this list of conditions and the following disclaimer
 *   in the documentation and/or other materials provided with the
 *   distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

#if !defined( CONFIG_H )
#define CONFIG_H

#include "../inc/Component.hpp"
#include "../inc/Lexical.hpp"

#define VERSION "Middlenebel v0.1.3-alpha Nebel-Docker"
// -------------------------------------------------------
#define CONFIG_FILE_PROXY "./ConfigProxy.nebel"
#define CONFIG_FILE_NEBEL "./ConfigNebel.nebel"

// FRONT---web--------> WEBPORT [WebAngular ]---( liteServer / docker )
// FRONT---comands----> PROXYPORT [Proxy ] ----> NEBELPORT [NEBELDOCKER]
// ---------------------------------- ATTRIBUTES ---------
#define ATT_DISABLE_PARSE           "disableParse"
#define ATT_DISABLE_PROXY           "disableProxy"
#define ATT_USE_CONFIG              "useConfig"
#define ATT_DUMP_CONFIG             "dumpConfig"
#define ATT_FRONT                   "front"
#define ATT_WEBPORT                 "webPort"
#define ATT_PROXYPORT               "proxyPort"
#define ATT_NEBELPORT               "nebelPort"
#define ATT_FRONTNAME               "frontName"
// ----------------------------- CFG DEFAULT VALUES ------
#define CFG_TRUE                    "true"
#define CFG_DEFAULT                 "Default"
#define CFG_ON                      "on"
#define CFG_OFF                     "off"
#define CFG_INTERNAL                "internal"
#define CFG_LITE                    "lite"
#define CFG_DOCKER                  "docker"
#define CFG_FRONTNAME               "my-angular-app"
#define CFG_WEBPORT                 "3000"
#define CFG_PROXYPORT               "8081"
#define CFG_NEBELPORT               "8080"

// -------------------------------------------------------
#define IS_CONFIG( A, B) (config->cfg(A)==B)
#define IS_NOT_CONFIG( A, B) (config->cfg(A)!=B)

// --- PARSER
#define TK_CONFIG "Config"

#define DUMP_CONFIGURATION { \
    cout << "Using: "<<cfg(ATT_USE_CONFIG)<<"\n"; \
    cout << "Configurations:\n"; \
    for (std::map<string,Config*>::iterator cfgItem = configurations.begin(); cfgItem != configurations.end(); ++cfgItem){ \
        std::cout << "\t" << (cfgItem->first) << "\n"; \
        for (std::map<string,string>::iterator attItem = cfgItem->second->attributes.begin(); attItem != cfgItem->second->attributes.end(); ++attItem){ \
            if ((attItem->first!=ATT_CLASSNAME)&&(attItem->first!=ATT_NAME)) \
                std::cout << "\t\t" << (attItem->first) << " : " << (attItem->second) << "\n"; \
        } \
    } \
}

class Config : public Component {
    private:

        void init();
        std::map<string, Config*> configurations;
        Config* config; // Can be use Config like Core
        Config(Lexical*, Component*);
    public:
        string fileName;
        string newFileName;

        Config(Component*);
        void loadConfig(string file);
        void saveConfig(string file, string script);

        void parse();
        void parseConfig();

        string cfg(string key, string defValue = "");

        virtual int getObjectNum();
        virtual string doQuit();
};
#endif // !defined( CONFIG_H )